Qt5: SIGINT kills just the mpl window and not the process itself

vdrhtc opened the pull request:
## PR Summary

This pull request fixes issue #13302 using `signal.set_wakeup_fd()`. 

The implemented `test_fig_sigint()` uses the example from the aforementioned issue. It's rather an integration test, though. It does not fail with the previous behaviour, but never passes either hanging the whole testing process.

I also changed `test_fig_signals()` to pass with the new code.

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

joelfrederico commented:
I'd downvote. Timers kill performance, they're an exceptionally bad idea. If you really want Python to handle a signal, you should register a handler with Qt: http://doc.qt.io/qt-5/unix-signals.html. (There is an analogue Windows method to this as well.)

There is some discussion as to whether this is a good idea or not: https://github.com/matplotlib/matplotlib/issues/13302#issuecomment-458184723

vdrhtc commented:
Removed QTimer, using `signal.set_wakeup_fd()` instead. Only tested on Linux!

KeyboardInterrupt is raised after `qApp.exec_()` when exit is due to caught SIGINT. Should be fine for everyone, I think, and be the default behaviour.

joelfrederico commented on lib/matplotlib/backends/backend_qt5.py:
```diff
@@ -1123,27 +1160,25 @@ class _BackendQT5(_Backend):
-                          # by _BackendQT5.interrupt_handler
-        timer.timeout.connect(lambda: None)
+
+        SIGINTHandler(qApp)
```
This may theoretically be garbage-collected immediately. Better to use a context manager.

vdrhtc replied:
I guess it wasn't because it was a QObj had a parent qApp, but now it's not relevant

joelfrederico commented on lib/matplotlib/backends/backend_qt5.py:
```diff
@@ -1123,27 +1160,25 @@ class _BackendQT5(_Backend):
+        signal.signal(signal.SIGINT, lambda sig, _: interrupt_handler())
 
         try:
             qApp.exec_()
```
Preferably, the context manager would wrap this line.

vdrhtc replied:
Done

joelfrederico commented on lib/matplotlib/tests/test_backend_qt.py:
```diff
@@ -139,7 +139,7 @@ def update(frame):
         else:
             # print(signal.getsignal(signal.SIGINT))
```
Hmmm... I left a print comment in here... bad form!

vdrhtc replied:
Removed this comment

vdrhtc replied:
And it was mine, btw =)

joelfrederico commented on lib/matplotlib/tests/test_backend_qt.py:
```diff
@@ -139,7 +139,7 @@ def update(frame):
             # print(signal.getsignal(signal.SIGINT))
             os.kill(os.getpid(), signal.SIGINT)
-            print("Simulated interrupt")
+            # print("Simulated interrupt")
```
Remove print comment

vdrhtc commented:
Current approach seems to work on Windows almost out of the box, only had to change the socket type to default (STREAM) since Windows didn't allow DGRAM. The test however now doesn't work since os.kill(os.getpid(), signal.SIGINT) kills CPython with pytest and everything.

I again faced (and captured) the bug which is about `qApp.exit()` not being called after the sole plot window is closed. Is this connected with %pylab qt5?

vdrhtc commented:
@tacaswell, it's cetainly not ready yet, I need to test more and try to simplify =)

@joelfrederico, if I understand everything correctly, Python registers some low-level handler for SIGINTS that is actually setting a flag for the VM and writing to the file descriptor according to `signal.set_wakeup_fd()` independently of what the code is doing at the high level (this is undocumented). So when we make a Qt class listening to socketpair's other end, it should react and reacts (from what I see) immediately, even when not running the Python interpreter. 

I also think I grasped what is in your zmq link with kernel32.SetConsoleCtrlHandler, I think it's also a way to try.

vdrhtc commented:
Sorry, messed up with commits and the author names a bit, it was all me. 

I've implemented the context manager (thanks, @joelfrederico), fixed the tests.

I think I am starting to like this code, but now I get 
`QSocketNotifier: Invalid socket 75 and type 'Read', disabling...`
in the jupyter notebook terminal stderr. Probably due to the TCP socket type.

joelfrederico commented:
So... yes, Python actually has a SIGINT handler that, in an undocumented manner, records that a SIGINT has occurred. That's all the handler does.

Then, before every python instruction, python checks to see if a SIGINT has occurred. If it has, the python code runs the current Python-registered signal handler function.

But what actually happens if you're in the C part of Qt when a signal happens? The handler records that a SIGINT has occurred and returns. That's it. The Python-registered signal handler function will happen the next time any Python code runs. But what happens if no Python code ever runs again? Qt will continue along as if nothing happened. It will never call the Python-registered signal handler function, so there will be no socket event, so it will never quit.

Instead, what we probably want to happen is to register our own low-level signal handler. Then, if a signal happens during Qt C, our low-level signal handler runs. It then writes to a socket and returns. Our Qt should have a QSocketNotifier listening on that socket. Once the Qt event loops gets around to checking that socket, it sees the write and then emits a signal that it has detected an interrupt. Up until this point, everything is in C, no Python.

The question is, what do we want the SIGINT slots to do? We have options. I think a solid option is to simply call the custom handler (that is effectively a closure) that you designed. That custom handler records that a SIGINT has happened, followed by a call to Qt `exit()`. That way, whether a SIGINT happens in Qt or in Python, the result is the same. Then the Qt event loop continues until it checks to see if `exit()` has been called. It then exits and returns from the place `exec_()` was called. Your closure had registered that a SIGINT had happened, so then we call the previous signal handler.

joelfrederico commented:
The astute observer will notice the problem: what happens with us overriding Python's SIGINT handler? Does it get reset at some point, or does ours take precedence? If so, we need to make sure it gets reset. I don't really know how to do this. It should be possible in theory, just a question of whether Python's low-level SIGINT handler is documented.

Like I said, I switched away from dealing with this, so I didn't tackle this issue.

vdrhtc commented:
@joelfrederico,
> But what actually happens if you're in the C part of Qt when a signal happens? The handler records that a SIGINT has occurred and returns. That's it. The Python-registered signal handler function will happen the next time any Python code runs. But what happens if no Python code ever runs again? Qt will continue along as if nothing happened. It will never call the Python-registered signal handler function, so there will be no socket event, so it will never quit.

https://github.com/python/cpython/blob/a1f9a3332bd4767e47013ea787022f06b6dbcbbd/Modules/signalmodule.c#L237-L252

Here I can vaguely see that the handler indeed sets the flag for the VM to check AND writes to wakeup fd after that. The `trip_signal()` func is called in the default low-level handler `signal_handler()` used then in `PyOS_setsig()`(defined [here](https://github.com/python/cpython/blob/a1f9a3332bd4767e47013ea787022f06b6dbcbbd/Python/pylifecycle.c#L2240)), so I guess it gets fully executed no matter what code is running, pure C library or CPython. This means that we can rely on that wakeup fd, as I did, and do not invent anything -- it looks like changing this low-level handler would be hard.

joelfrederico commented:
Is that file descriptor/socket documented to be used always and only for SIGINT? If so, great! Done.

If not, I'm personally very wary of relying on undocumented behavior.

vdrhtc commented:
@joelfrederico 
No, from the code it's for any signal -- it writes the signal number into the fd. Qt registers that write (from now on we don't care about the value written), and then calls Python. Python now can find it's flag, finds that was actually a SIGINT, and, finally, acts according to our set handler.

It's not fully documented, _when_ the descriptor is written to (I mean, at the same time the signal is caught by the process), but I think it's the only purpose of this function. I think I may ask about it in their repo, should I? Maybe they will document that more precisely, I can't imagine why they wouldn't want to.

joelfrederico commented:
Hey it *is* documented! Python function [`signal.set_wakeup_fd`](https://docs.python.org/3/library/signal.html#signal.set_wakeup_fd). That's awesome! I was rooting around in the C API and it was in the `signal` module the whole time!

Okay, so all you have to do is adapt your code to create a listening socket, pass that in to `signal.set_wakeup_fd` and into a QSocketNotifier, and you're good! Just reset it when you're done. The QSocketNotifier just needs to run some Python, which runs your signal handler and everything's fine.

I feel silly for not noticing this earlier. Well done!

vdrhtc commented:
> it was in the signal module the whole time!

Yes, that's what I what trying to convey, it was from the beginning all on StackOverflow as well =) But not in the first answer, so coding was much harder this time!

> I have no idea why PyZMQ gets this so wrong

Because they didn't have the function on Windows back in 2014 and earlier, I guess:
![default](https://user-images.githubusercontent.com/3819012/52082561-2530c080-25ae-11e9-8799-8c5b543ce040.png)

And now no one would want to change that =) 

Is it possible to set a handler like QSocketNotifier in ZMQ? I really don't know much about how it works

vdrhtc commented:
@joelfrederico, sorry, I wasn't precise: I said they didn't have it in 2014, but that's not true. The thing is that `signal.set_wakeup_fd` started supporting socket handles _only since Python 3.5_ (released on Fall, 2015), as said in the doc:
![default](https://user-images.githubusercontent.com/3819012/52107517-b7b17e00-2607-11e9-87bd-ea921a2f7fd8.png)
Additionally, you could not create a socketpair in Windows before Python 3.5:
![default](https://user-images.githubusercontent.com/3819012/52107489-99e41900-2607-11e9-837a-91eda5e38064.png)
So I guess these problems made them to create their own win32 utility in 2014 when Python 3.5 has not yet been released.

But _now_ the approach works as expected, as I have tested -- the sigint test passes on Windows and on Linux.

jklymak commented:
I'm going to close this as abandoned, but please feel free to reopen if anyone wants to pursue it...

QuLogic commented:
I'm not sure what you mean; this seems to have been updated to match the latest request. It needs a review now, but I tested it and it works.

QuLogic commented:
OK, it also needs a rebase, but since none of us did any review, I can imagine that was not noticed.

@vdrhtc please rebase this, and maybe squash your older attempts away.

jklymak commented:
I mean the last request was 1.5 y ago and no one came back to it.  Apparently not urgent.  But if you feel its useful, feel free to reopen!

vdrhtc commented:
I hope everything is all right with the squash.

Yes, I am still waiting for your approval of this pull request, and I still do believe this functionality will be useful for many people around who does not want to mess with Qt internals themselves.

vdrhtc commented:
But, yes, 1.5 year break is a bit too long, as I can tell, cause it's really hard to recall the details now. I find that I also have some other branches with small fixes in my fork, but I have completely forgotten what they were aimed at.

QuLogic commented on lib/matplotlib/backends/backend_qt5.py:
```diff
@@ -17,10 +18,19 @@
+    QT_API,
+    _devicePixelRatioF,
+    _isdeleted,
+    _setDevicePixelRatioF,
```
Please revert this change, i.e., `_getSaveFileName` is not used, and the restyling is unnecessary churn.

QuLogic commented on lib/matplotlib/backends/backend_qt5.py:
```diff
@@ -17,10 +18,19 @@
+    _setDevicePixelRatioF,
 )
 
+from PyQt5 import QtNetwork
```
Have you tried this with PySide? I guess this needs to go in `qt_compat` as well.

vdrhtc replied:
Yes, it certainly should. I have modified the qt_compat and have tested with PySide2, it works OK

tacaswell replied:
We can't have pyqt5 imports in this file, all of that needs to happen in `qt_compat` and shimmed across pyqt4 / pyside / pyqt5 / pyside2

QuLogic commented on lib/matplotlib/backends/backend_qt5.py:
```diff
@@ -1019,6 +1031,80 @@ def trigger(self, *args, **kwargs):
 
 
+@cbook.deprecated("3.0")
+def error_msg_qt(msg, parent=None):
```
Remove.

QuLogic commented on lib/matplotlib/backends/backend_qt5.py:
```diff
@@ -1019,6 +1031,80 @@ def trigger(self, *args, **kwargs):
+
+
+@cbook.deprecated("3.0")
+def exception_handler(type, value, tb):
```
Remove.

tacaswell replied:
I think this function is being accidentally resurrected?

QuLogic commented on lib/matplotlib/backends/backend_qt5.py:
```diff
@@ -1019,6 +1031,80 @@ def trigger(self, *args, **kwargs):
+    if len(msg):
+        error_msg_qt(msg)
+
+class allow_interrupt():
```
Make non-public:
```suggestion
class _allow_interrupt:
```

QuLogic commented on lib/matplotlib/tests/test_backend_qt.py:
```diff
@@ -157,6 +216,7 @@ def test_dpi_ratio_change():
     Make sure that if _dpi_ratio changes, the figure dpi changes but the
     widget remains the same physical size.
     """
+    return
```
This doesn't look necessary.

vdrhtc replied:
Ah yes, this test fails on my notebook as the Qt does not have enough time to open the window. Its not my fault, I guess, and I can suggest an easy fix for that. It will be included in the next commit.

QuLogic commented on lib/matplotlib/tests/test_backend_qt.py:
```diff
@@ -76,15 +134,16 @@ def CustomHandler(signum, frame):
     signal.signal(signal.SIGINT, CustomHandler)
 
-    # mainloop() sets SIGINT, starts Qt event loop (which triggers timer and
+    # mainloop() sets SIGINT, starts Qt event loop (which trigers timer and
```
Revert typo.

QuLogic commented on lib/matplotlib/tests/test_backend_qt.py:
```diff
@@ -47,8 +65,48 @@ def test_fig_close(backend):
+    except KeyboardInterrupt as e:
+        assert len(Gcf.get_all_fig_managers()) == windows_left
+        return
+    assert False
```
The test won't pass with this.

vdrhtc replied:
With line 102 and no pyplot.close() in the handler which we can't have there since we can't import pyplot in backend_qt5, yes, it won't. Otherwise its OK?

QuLogic replied:
Oh, I missed the early return; please add an explicit message on the assertion then.

QuLogic commented on lib/matplotlib/backends/backend_qt5.py:
```diff
@@ -1019,6 +1031,80 @@ def trigger(self, *args, **kwargs):
+            self.old_sigint_handler()
+
+    def _readSignal(self):
+        # Read the written byte.
+        # Note: readyRead is blocked from occurring again until readData()
+        # was called, so call it, even if you don't need the value.
```
Not sure what this comment is getting at.

vdrhtc replied:
I can't elaborate, this is not my comment. It belongs to the author who proposed [this solution](https://stackoverflow.com/questions/4938723/what-is-the-correct-way-to-make-my-pyqt-application-quit-when-killed-from-the-co). The tests pass even without the qt_socket.readData(1). I would suggest leaving as is, so the next person will be able to look deeper in the consequences of the stated fact.

vdrhtc replied:
Or maybe replace all the body of the function with a pass statement?

tacaswell replied:
This is one of those self-closing box situations, we are using this to wake up so we push something into the file, but it won't signal it is ready to be "woken up" again until we fully drain it.

If we only got sigint we wouldn't need this, but have to "re-arm" after other signals.

QuLogic commented on lib/matplotlib/backends/backend_qt5.py:
```diff
@@ -3,12 +3,13 @@
 import matplotlib
 
-from matplotlib import backend_tools, cbook
+from matplotlib import backend_tools, cbook, pyplot
```
I don't think we want a `pyplot` import leaking into backends, as that will cause many things to be set that are not necessarily desired.

vdrhtc replied:
OK, I have removed it. It was only used to close the figure before app.quit(), and that was used to check that there are no more open windows. I have modified the test so it does not care how many figure windows are left.

tacaswell commented on lib/matplotlib/backends/backend_qt5.py:
```diff
@@ -1019,6 +1031,80 @@ def trigger(self, *args, **kwargs):
+    def _handle(self, *args):
+        self.interrupt_caught = True
+        if self.interrupted_qobject is qApp:
+            pyplot.close()
```
I am not sure we want this.  This will only close the current figure (not all of them).  We should either just leave the windows be (assuming that something outside of us will spin the event loop or not as they want) or do `plt.close('all')` 

I strongly prefer we leave the figures alone here.

tacaswell commented:
@joelfrederico described the problem perfectly in https://github.com/matplotlib/matplotlib/pull/13306#issuecomment-459128896 

I think there are three questions here:

1. is this the right implementation of how to shuffle the SIGINT back to Python and cause it to exit the Qt main loop rather exiting the process
2. do we want to ship this code in Matplotlib
3. should it be enabled by default

For 1: I don't think we are there yet quite yet due to the `QNetwork` import.  I have a slightly different implementation of this in one of my DAMA projects (https://github.com/bluesky/bluesky/blob/cde1d87294a41d48d0a92d6cae0ce9c508e20388/bluesky/utils.py#L1384-L1495 that in light of some of this discussion suggests we may be able to simplify the windows version!).  Can we use the QSocketNotifier approach here?  I also think that this context manager should go in qt_compat (just for organizational reasons).

For 2: maybe?  This is a rather nasty bit of code to carry around and it is a bit out of our core mission, but given that there are no at least 2 contexts where this was needed maybe we should?  Does qtpy have something for this?

For 3: I'm leaning to "yes", but giving people a knob to turn it off.  As a quick test the behavior with tk is to raise `KeyboardInterpt` not exit.



---

Reminding my self what is going on here, I see I tried to sell @vdrhtc on code from my other life doing data acquisition, how did that go?

vdrhtc commented on lib/matplotlib/backends/backend_qt5.py:
```diff
@@ -145,7 +145,7 @@ def _allow_super_init(__init__):
 
     else:
-        # To work around lack of cooperative inheritance in PyQt4, PySide,
+        # To work around lack of cooperative inheritance in PyQQtNetworkt4, PySide,
```
This was unintentional, sorry, will fix later

vdrhtc commented:
@tacaswell 
Hi! Yes, I was looking into bluesky for some time, and I got some thoughts against fully investing myself into it. the main problem is that we would have to merge our code with yours, the latter being the sweep and dynamic plotting engine, while the former would contain the high-level logic of the experiments (which can become quite involved).  And from all my experience, I don't believe that a seamless connection will be fast and easy. There is not magic pill to cure all diseases. Moreover, existence of several alternative approaches to the same problem means decentralization and diversity, which I believe are not bad things, right?

Considering the nastyness of the code that I've found -  I don't think that ```qt_compat.py``` iteslf is much less nasty. But it does not grow much with time, so I thing we excuse that. By the way, optimization with non-native libraries will always cause nasty code, so that's not about bad coding skills or wrong architecture, we just don't have any other choice (except cutting the functionality). 

For 1: I am sorry, I don't understand what you are asking, could you clarify?

tacaswell commented on lib/matplotlib/tests/test_backend_qt.py:
```diff
@@ -47,8 +65,48 @@ def test_fig_close(backend):
 
 
+@pytest.mark.backend('Qt5Agg')
+@pytest.mark.parametrize("target, kwargs, windows_left", [
```
windows left is never checked.

tacaswell commented on lib/matplotlib/tests/test_backend_qt.py:
```diff
@@ -47,8 +65,48 @@ def test_fig_close(backend):
+        else:
+            return ln,
+
+    ani = FuncAnimation(fig, update, frames=range(0, 100),
```
This is a super clever way of injecting a SIGINT!  However, it may be simpler to use a single-shot timer.

tacaswell commented:
Trying to work through the history of this, it looks like back on 2012 ( df0a11269f7b92173c20f8de4a95f9d4c4891a4c /  #851 ) we added

```python
signal.signal(signal.SIGINT, signal.SIG_DFL)
```

(which was moved around a bunch and improved a bit in #13027, #16311, #16330) however I think the core of the functionality was broken by pyqt5.5 (https://www.riverbankcomputing.com/static/Docs/PyQt5/incompatibilities.html#pyqt-v5-5) which changed from just printing out exceptions to exiting the process if exceptions were raised from python run.  Unfortunately getting a version of pyqt5 5.4 that runs on a modern Python seems like less than trivial and I am not sure it is worth the effort to sort out exactly when this stopped working :-p 

@vdrhtc For question 1, the question is "this looks complex, is this really the best way to get ctrl-c promptly?"

tacaswell commented on lib/matplotlib/tests/test_backend_qt.py:
```diff
@@ -176,6 +236,7 @@ def test_dpi_ratio_change():
         size = qt_canvas.size()
 
         qt_canvas.manager.show()
+        plt.pause(0.5)  # otherwise the window does not have enough time to open, and the test fails
```
Ah great!  This had been bugging me ....

tacaswell commented on lib/matplotlib/backends/qt_compat.py:
```diff
@@ -216,3 +220,54 @@ def _setDevicePixelRatioF(obj, val):
+        self.old_sigint_handler = None
+        self.interrupt_caught = None
+
+        QAS = QtNetwork.QAbstractSocket
```
Can / should we defer setting up the sockets until we know that we can safely register a signal handler?  In the case of julia binding into Python, the julia side handles all of the signals so maybe we should do nothing and leave this all to them?

tacaswell replied:
This comment was handled by protecting the call to the context manager and with the check at the top of the init.

tacaswell commented:
@stevengj Would you be willing to confirm that this branch does not break your usage in julia?

tacaswell commented:
@vdrhtc I took the liberty of pushing a bunch of commits to this branch (as just pushing the commits was less work than trying to describe in English what I thought had to be done).  The main things were:
 
 - mechanical style cleanups
 - making sure we respect the `None` signal handler
 - passing the values passed to the handler back so the original handler still runs
 - be a bit more paranoid about checking what handler _was_ registered.

Do we also want to use something like https://github.com/bluesky/bluesky/blob/9137603d942a052027d310b5c12204e3883344d8/bluesky/utils.py#L1492-L1515 to really make sure that nothing escapes to kill the process?  I am inclined to say no because we already suppress exceptions in our callbacks.  If the user writes their own code that can run in the c++ event loop and raise an exception, that is on them to make sure they have covered their bases.

tacaswell commented on lib/matplotlib/backends/qt_compat.py:
```diff
@@ -216,3 +220,57 @@ def _setDevicePixelRatioF(obj, val):
+        self.wsock.setblocking(False)
+        # First Python code executed gets any exception from
+        # the signal handler, so add a dummy handler first
+        self.qt_socket.readyRead.connect(lambda: None)
```
I don't think we actually need this.  Because we are installing our own signal handler the `KeyboardInterupt` does not get injected so we don't have this problem.

tacaswell commented:
In summary, I took another pass at this, I think it is heading in the right direction, but have some questions / concerns.

vdrhtc commented:
@tacaswell Thank you very much for looking into the code! I have made some modifications according to your comments, please, see below

jklymak commented:
Moving the milestone on this, but feel free to move back.  @vdrhtc please don't be shy about re-pinging folks.  PRs definitely get buried, so sorry if this is not giving along expeditiously.

vdrhtc commented:
@tacaswell @jklymak Hi, maybe I should rebase again?

tacaswell commented:
@vdrhtc Can you rebase again?  I think this is close.

tacaswell commented on lib/matplotlib/backends/qt_compat.py:
```diff
@@ -216,3 +220,48 @@ def _setDevicePixelRatioF(obj, val):
+        self.interrupted_qobject = qApp
+        self.old_fd = None
+        if old_sigint_handler in (None, signal.SIG_IGN, signal.SIG_DFL):
+            raise ValueError(f"Overriding a SIGINT handler {old_sigint_handler} is not allowed")
```
I think this is the line that flake8 is unhappy about?

vdrhtc commented:
@tacaswell @jklymak rebased, please, see again

tacaswell commented:
@vdrhtc Sorry this fell off my radar (again).  I took the liberty of rebasing this to move it forward, I hope you do not mind.

vdrhtc commented:
@tacaswell, that's very nice, thank you!

tacaswell commented on lib/matplotlib/backends/qt_compat.py:
```diff
@@ -216,3 +220,49 @@ def _setDevicePixelRatio(obj, val):
+    def __init__(self, qApp, old_sigint_handler):
+        self.interrupted_qobject = qApp
+        self.old_fd = None
+        if old_sigint_handler in (None, signal.SIG_IGN, signal.SIG_DFL):
+            raise ValueError(f"Old SIGINT handler {old_sigint_handler}"
+                             f" will not be overridden")
```
After a conversation with @QuLogic on the phone I agree.  Having the enter/exit be no-ops and keeping all of this logic internal seems better.

Maybe we should rename the context manager to be `_maybe_allow_interput`? 

I may have time to push a commit to do this in the next few days...

anntzer commented:
Given that @tacaswell asked me to review this... is there a reason not to use QSocketNotifier instead?  (basically the approach at https://doc.qt.io/qt-5/unix-signals.html, which we can just link to for reference, and also the one @tacaswell linked to in bluesky)
This way we don't have to set up network-related stuff (I know nothing of TCP), nor fiddle with qt_compat to import QtNetwork.

(I only skimmed through the discussion, apologies if this was discussed above and I missed the discussion.)

tacaswell commented:
My memory of this is that there are issues with QSocketNotifier on windows.

QuLogic commented:
I looked around and see that `QSocketNotifier` only works with Winsock2 sockets on Windows (https://stackoverflow.com/questions/39327570/qsocketnotifier-does-not-work-on-windows-for-serialport-reading); I'm not sure if that means it won't work with the BSD-style sockets created by `socketpair`.

anntzer commented on lib/matplotlib/backends/qt_compat.py:
```diff
@@ -216,3 +220,49 @@ def _setDevicePixelRatio(obj, val):
+    def __init__(self, qApp, old_sigint_handler):
+        self.interrupted_qobject = qApp
+        self.old_fd = None
+        if old_sigint_handler in (None, signal.SIG_IGN, signal.SIG_DFL):
```
1) why do you want to not override SIG_DFL?
2) the first time you call signal.getsignal(SIGINT) (having not overriddent SIGINT before), you actually don't get SIG_DFL, but the private _signal.default_int_handler.  Perhaps that's an oversight in the docs, but in any case that case needs to be handled too (or SIG_DFL also needs to be overridable).

vdrhtc replied:
Sorry for not answering for so long! This is the reason: https://github.com/matplotlib/matplotlib/pull/16311. SIG_DFL is used to indicate that a non-Python signal handler was installed for SIGINT. We have discussed with @tacaswell that this PR should not break #16311 and thus do not force new logic in that case.

anntzer replied:
> SIG_DFL is used to indicate that a non-Python signal handler was installed for SIGINT.

I am confused: the docs (https://docs.python.org/3/library/signal.html#signal.getsignal) "signal.SIG_DFL means that the default way of handling the signal was previously in use, and None means that the previous signal handler was not installed from Python."

vdrhtc replied:
Oh yes, I have confused that with the None case. I guess that is the ''knob'' to turn off the logic that @tacaswell has left. https://github.com/matplotlib/matplotlib/pull/13306#issuecomment-672509173 shows that, probably. So you first set up the DFL signal by
```Python
signal.signal(signal.SIGINT, signal.SIG_DFL)
```
and then the previous logic is restored.

anntzer commented:
The relatively simpler(?)
```python
@_Backend.export
class _BackendQT5(_Backend):
    FigureCanvas = FigureCanvasQT
    FigureManager = FigureManagerQT

    @staticmethod
    def mainloop():
        with _maybe_allow_interrupt(qApp):
            qApp.exec_()


@contextlib.contextmanager
def _maybe_allow_interrupt(qapp):
    signal.signal(signal.SIGINT, lambda *args: qapp.closeAllWindows())  # Needs restoring at exit, etc.
    wsock, rsock = socket.socketpair()
    wsock.setblocking(False)
    old_wakeup_fd = signal.set_wakeup_fd(wsock.fileno())
    sn = QtCore.QSocketNotifier(rsock.fileno(), QtCore.QSocketNotifier.Read)

    @sn.activated.connect
    def on_signal(*args):
        sn.setEnabled(False)
        qapp.processEvents()

    try:
        yield
    finally:
        signal.set_wakeup_fd(old_wakeup_fd)
        wsock.close()
        rsock.close()
```
works for me both on Windows and on Linux.

jklymak commented:
@vdrhtc Please see @anntzer comment above...

vdrhtc commented:
@jklymak sorry for the delay, I will look into that on the next week!

vdrhtc commented:
@anntzer thank you very much, this looks like a very clear and concise implementation! I would like to incorporate that. As far as I can tell, you do not check for the None, SIGIGN and SIG_DFL signals anymore? Should I add them? (please, see my [answer here](https://github.com/matplotlib/matplotlib/pull/13306#discussion_r629969788) as well)

Does the QtCore.QSocketNotifier read the data from the socket upon noticing its availability? If we accidentally get a signal different from SIGINT first, won't that block the context manager from processing all signals afterwards? I also read that ```A disabled notifier ignores socket events (the same effect as not creating the socket notifier).``` so it rather looks like a deliberately single-use handler? See https://github.com/matplotlib/matplotlib/pull/13306#discussion_r460279273

Also, it seems to me that you do not call the old handler with the supplied args here, so no KeyboardInterrupt will be raised, but that was the previous behavior before the break by pyqt5.5, so I personally would prefer to leave that call as before. It allows the wrapper program to understand that an interrupt actually happened.

anntzer commented:
My example implementation was just an example, not a complete thing (hence the lack of using the old handlers, etc.)  To be honest I forgot what parts I didn't implement just out of laziness and what parts are actually missing, so you'll have to work that out yourself :-) (or rather, it'll take me some time to go though all the docs again to figure that out...)

jklymak commented:
@vdrhtc I moved this to "Draft" to keep it off our radar until you were done with it.  However, if you are done, feel free to move to "Ready for review".  Thanks!

anntzer commented:
Can you describe how to repro the crash you mention?

vdrhtc commented:
@anntzer, the way is to replace this [line](https://github.com/vdrhtc/matplotlib/blob/6d2ea0961efdc879ec17019af4a022ae907d08a3/lib/matplotlib/backends/qt_compat.py#L280) by a ```pass``` statement so the socket is not read from.

tacaswell commented:
@vdrhtc I'm going to do this rebase and take over this PR, sorry this has sat for so long.

tacaswell commented:
I am now concerned why this _did_ work for me locally before changing out the application for the loop in pause.  My guess is something in detailed qt versions that I do not want to chase down.

These tests also pass for me locally.

QuLogic commented:
They also pass on CI, so is all fine now?

tacaswell commented:
> They also pass on CI, so is all fine now?

Yes I think so.

QuLogic commented on lib/matplotlib/backends/qt_compat.py:
```diff
@@ -191,3 +194,64 @@ def _setDevicePixelRatio(obj, val):
+            rsock.fileno(), _enum('QtCore.QSocketNotifier.Type').Read
+        )
+
+        @sn.activated.connect
+        def on_signal(*args):
+            rsock.recv(
+                sys.getsizeof(int))  # clear the socket to re-arm the notifier
```
`set_wakeup_fd` writes a single byte, not an `int`. Surprising this doesn't block waiting for the remaining bytes.

QuLogic replied:
So seems that this could be shorter:
```suggestion
        # Clear the socket to re-arm the notifier.
        sn.activated.connect(lambda *args: rsock.recv(1))
```

anntzer commented on lib/matplotlib/backends/qt_compat.py:
```diff
@@ -74,32 +77,33 @@
 
 def _setup_pyqt5plus():
-    global QtCore, QtGui, QtWidgets, __version__, is_pyqt5, \
+    global QtCore, QtGui, QtWidgets, QtNetwork, __version__, is_pyqt5, \
```
I don't think we need QtNetwork anymore (and can thus revert a few changes here).

QuLogic replied:
#20882

anntzer commented on lib/matplotlib/tests/test_backend_qt.py:
```diff
@@ -33,8 +33,22 @@ def qt_core(request):
+    from functools import partial
+
+    if hasattr(signal, "CTRL_C_EVENT"):
+        from win32api import GenerateConsoleCtrlEvent
```
this adds a dependency on pywin32 for tests on Windows?  (that's fine, but needs to be documented or at least made an importorskip or whatnot?)

QuLogic replied:
It looks like we install that for some other reason on appveyor, and it gets pulled in through something else in Azure.

QuLogic replied:
#20883

QuLogic commented:
Oops, crossed with your review.

anntzer commented:
my fault, shouldn't have decided to review after all :)
